homeassistant/components/device_tracker/icloud.py

```diff
@@ -1,100 +1,449 @@
+    hass.services.register(DOMAIN, 'icloud_reset_account',
+                           reset_account_icloud, schema=SERVICE_SCHEMA)
+
+    def setinterval(call):
+        """Call the update function of an icloud account."""
+        accounts = call.data.get(ATTR_ACCOUNTNAME, ICLOUDTRACKERS)
+        interval = call.data.get(ATTR_INTERVAL)
+        devicename = call.data.get(ATTR_DEVICENAME)
+        for account in accounts:
+            if account in ICLOUDTRACKERS:
+                ICLOUDTRACKERS[account].setinterval(interval, devicename)
+
+    hass.services.register(DOMAIN, 'icloud_set_interval', setinterval,
+                           schema=SERVICE_SCHEMA)
+
+    # Tells the bootstrapper that the component was successfully initialized
+    return True
+
+
+class Icloud(object):  # pylint: disable=too-many-instance-attributes
+    """Represent an icloud account in Home Assistant."""
+
+    def __init__(self, hass, username, password, name, googletraveltime, see):
+        """Initialize an iCloud account."""
+        # pylint: disable=too-many-arguments,too-many-branches
+        # pylint: disable=too-many-statements,too-many-locals
+        self.hass = hass
+        self.username = username
+        self.password = password
+        self.api = None
+        self.accountname = name
+        self.devices = {}
+        self.seen_devices = {}
+        self.googletraveltime = googletraveltime
+        self._overridestates = {}
+        self._intervals = {}
+        self.see = see
+
+        self._trusted_device = None
+        self._verification_code = None
+
+        self._attrs = {}
+        self._attrs[ATTR_ACCOUNTNAME] = name
+
+        self.reset_account_icloud()
+
+        randomseconds = random.randint(10, 59)
+        track_utc_time_change(
+            self.hass, self.keep_alive,
+            second=randomseconds
+        )
+
+    def reset_account_icloud(self):
+        """Reset an icloud account."""
+        from pyicloud import PyiCloudService
+        from pyicloud.exceptions import PyiCloudFailedLoginException
+        try:
+            icloud_dir = self.hass.config.path('icloud')
+            if not os.path.exists(icloud_dir):
+                os.makedirs(icloud_dir)
+            self.api = PyiCloudService(
+                self.username, self.password,
+                cookie_directory=icloud_dir,
+                verify=True)
+            self.devices = {}
+            self._overridestates = {}
+            self._intervals = {}
+            for device in self.api.devices:
+                status = device.status(DEVICESTATUSSET)
+                devicename = slugify(status['name'].replace(' ', '', 99))
+                if devicename not in self.devices:
+                    self.devices[devicename] = device
+                    self._intervals[devicename] = 1
+                    self._overridestates[devicename] = None
+                    self.update_device(devicename)
+
+        except PyiCloudFailedLoginException as error:
```

Comment:
Wrap the `try…except` around the exact code that can raise, leave the rest out of it.